### PR TITLE
ovl_En_Nimotsu OK and documented

### PR DESCRIPTION
**Pull request by StickyThwomp**
Hey all. First attempt at contribution here. I *think* I've addressed everything that should be addressed, but please let me know anything I might be missing.

Of specific concern is whether more can/should be done surrounding:
- extern UNK_TYPE D_06013380;
    - I was unable to follow what type that should be declared as, or any name/purpose to assign to it, so I'm open to suggestions/guidance on how to tackle that.
- The last two members of the actor struct:
   - recognized one as a coundown to something, but couldn't quite determine what the coundown was to (in game terms) nor what the flag member was or what purpose it served. (Is there an easy-ish way to identify if something else in the game would be looking for that to help determine it's purpose?

Edit: Also, stealing reservation from MrPolymorph... Appeared abandoned.
Edit2: All questions above answered thanks to a bit of help!

Thanks!
-StickyThwomp

### CODE REVIEW

**Comment by hensldm** [comment]
Not sure if your questions got answered but 
1. That particular symbol belongs to the actor's object file and is addressed in https://github.com/zeldaret/mm/blob/master/docs/tutorial/data.md#segmented-pointers-and-object-symbols and objects in general in https://github.com/zeldaret/mm/blob/master/docs/tutorial/object_decomp.md and the corresponding example tutorial.
2. There is not really an easyish way. Most people use a combination of reading the code/using the practice rom (KZ) to help document what the struct members do. To know if another actor might interact with it is mainly up to game knoweldge, though I'm pretty sure most actors are pretty independent and rarely interact in such ways, though it is possible of course.

**Comment by StickyThwomp** [comment]
Thanks hensldm, I'll take a look at those object links and see if I can get a bit more information out of this.

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 1]
```diff
@@ -28,25 +31,84 @@ const ActorInit En_Nimotsu_InitVars = {
...
+    }
 
-extern UNK_TYPE D_06013380;
+    EnNimotsu_UpdateCollision(this, play);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Nimotsu/func_80BE1C80.s")
+void EnNimotsu_Draw(Actor* thisx, PlayState* play) {
+    EnNimotsu* this = THIS;
+    u32 pad;
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c:94]:
```suggestion
    s32 pad;
    EnNimotsu* this = THIS;
```

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 2]
```diff
@@ -28,25 +31,84 @@ const ActorInit En_Nimotsu_InitVars = {
...
+    position.y = this->actor.world.pos.y;
+    position.z = this->actor.world.pos.z + 2.0f;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Nimotsu/EnNimotsu_Update.s")
+    scale.x = 0.2f;
+    scale.y = 0.2f;
+    scale.z = 0.2f;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Nimotsu/EnNimotsu_Draw.s")
+    func_800BC620(&position, &scale, 0xFFU, play);
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c:112]:
```suggestion
    func_800BC620(&position, &scale, 0xFF, play);
```

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.h [hunk 3]
```diff
@@ -7,8 +7,11 @@ struct EnNimotsu;
 
 typedef struct EnNimotsu {
     /* 0x000 */ Actor actor;
-    /* 0x144 */ char unk_144[0x98];
-} EnNimotsu; // size = 0x1DC
+    /* 0x144 */ char unk_144[0x48];
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.h:10]:
```suggestion
    /* 0x144 */ UNK_TYPE1 unk_144[0x48];
```
`UNK_TYPE1` is preferred as the type for unknown struct padding.

**Review by hensldm — COMMENTED** [review]
Nice work on your first actor. Mostly just style comments.

File: assets/xml/objects/object_boj.xml [hunk 4]
```diff
@@ -87,8 +87,8 @@
         <Animation Name="object_boj_Anim_011F84" Offset="0x11F84" />
         <Animation Name="object_boj_Anim_0128F4" Offset="0x128F4" />
         <Animation Name="object_boj_Anim_012E84" Offset="0x12E84" />
-        <DList Name="object_boj_DL_013380" Offset="0x13380" />
-        <Texture Name="object_boj_TLUT_0134C0" OutName="tlut_0134C0" Format="rgba16" Width="16" Height="16" Offset="0x134C0" />
-        <Texture Name="object_boj_Tex_0136C0" OutName="tex_0136C0" Format="ci8" Width="16" Height="16" Offset="0x136C0" />
+        <DList Name="gBojBombShopBagDL" Offset="0x13380" />
+        <Texture Name="gBojBombShopBagTLUT" OutName="bomb_shop_bag_tlut" Format="rgba16" Width="16" Height="16" Offset="0x134C0" />
+        <Texture Name="gBojBombShopBagTEX" OutName="bomb_shop_bag_tex" Format="ci8" Width="16" Height="16" Offset="0x136C0" />
```
[the remark this answers is not in the document]
Comment (StickyThwomp) [reply assets/xml/objects/object_boj.xml]: Addressed in commit 137a1c8

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 5]
```diff
@@ -28,25 +31,84 @@ const ActorInit En_Nimotsu_InitVars = {
...
-    { COLTYPE_NONE, AT_NONE, AC_NONE, OC1_ON | OC1_TYPE_ALL, OC2_TYPE_2, COLSHAPE_CYLINDER, },
-    { ELEMTYPE_UNK4, { 0x00000000, 0x00, 0x00 }, { 0x00000000, 0x00, 0x00 }, TOUCH_NONE | TOUCH_SFX_NORMAL, BUMP_NONE, OCELEM_ON, },
+static ColliderCylinderInit sCylinderInit = {
+    { COLTYPE_NONE, AT_NONE, AC_NONE, OC1_ON | OC1_TYPE_ALL, OC2_TYPE_2, COLSHAPE_CYLINDER },
+    { ELEMTYPE_UNK4,
+      { 0x00000000, 0x00, 0x00 },
+      { 0x00000000, 0x00, 0x00 },
+      TOUCH_NONE | TOUCH_SFX_NORMAL,
+      BUMP_NONE,
+      OCELEM_ON },
```
[the remark this answers is not in the document]
Comment (StickyThwomp) [reply src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c]: Addressed in commit 137a1c8

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 6]
```diff
@@ -28,25 +31,84 @@ const ActorInit En_Nimotsu_InitVars = {
...
+
+    this->dustTimer = 10;
+    this->actor.gravity = -0.5f;
+
+    Actor_SetScale(&this->actor, 0.01f);
+}
+
+void EnNimotsu_Destroy(Actor* thisx, PlayState* play) {
+    EnNimotsu* this = THIS;
+    Collider_DestroyCylinder(play, &this->collider);
```
[the remark this answers is not in the document]
Comment (StickyThwomp) [reply src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c]: Addressed in commit 137a1c8

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 7]
```diff
@@ -28,25 +31,84 @@ const ActorInit En_Nimotsu_InitVars = {
...
+}
+
+void EnNimotsu_Update(Actor* thisx, PlayState* play) {
+    EnNimotsu* this = THIS;
+    u32 pad;
```
[the remark this answers is not in the document]
Comment (StickyThwomp) [reply src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c]: Addressed in commit 137a1c8

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 8]
```diff
@@ -28,25 +31,84 @@ const ActorInit En_Nimotsu_InitVars = {
...
+    Actor_MoveWithGravity(&this->actor);
+
+    if (!(this->dustDone & 1) && (this->actor.bgCheckFlags & 1)) {
+
+        if (DECR(this->dustTimer) == 0) {
+            this->dustDone |= 1;
+        }
 
-extern ColliderCylinderInit D_80BE1FB0;
+        if ((play->state.frames % 3U) == 0) {
```
[the remark this answers is not in the document]
Comment (StickyThwomp) [reply src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c]: Addressed in commit 137a1c8

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 9]
```diff
@@ -28,25 +31,84 @@ const ActorInit En_Nimotsu_InitVars = {
...
+            dustPosition.x = this->actor.world.pos.x + randPlusMinusPoint5Scaled(15.0f);
+            dustPosition.y = this->actor.world.pos.y;
+            dustPosition.z = this->actor.world.pos.z + randPlusMinusPoint5Scaled(15.0f);
+            Actor_SpawnFloorDustRing(play, &this->actor, &dustPosition, 20.0f, 0, 2.0f, (s16)0, (s16)0, (u8)0);
```
[the remark this answers is not in the document]
Comment (StickyThwomp) [reply src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c]: Addressed in commit 137a1c8

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c (hunk 1 above)
Comment (StickyThwomp) [reply src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c]: Addressed in commit 137a1c8

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c (hunk 2 above)
Comment (StickyThwomp) [reply src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c]: Addressed in commit 137a1c8

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.h (hunk 3 above)
Comment (StickyThwomp) [reply src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.h]: Addressed in commit 137a1c8

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 10]
```diff
@@ -28,25 +31,84 @@ const ActorInit En_Nimotsu_InitVars = {
...
+      BUMP_NONE,
+      OCELEM_ON },
     { 10, 30, 0, { 0, 0, 0 } },
 };
 
-#endif
+void EnNimotsu_UpdateCollision(EnNimotsu* this, PlayState* play) {
+    Collider_UpdateCylinder(&this->actor, &this->collider);
+    CollisionCheck_SetOC(play, &play->colChkCtx, &this->collider.base);
+    Actor_UpdateBgCheckInfo(play, &this->actor, 32.0f, 30.0f, 0.0f, 4U);
```
[the remark this answers is not in the document]
Comment (StickyThwomp) [reply src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c]: Addressed in commit 137a1c8

**Comment by isghj5** [comment]
You don't really need to reply to each requested change when you fix it, it's okay to just resolve the conversation.

**Comment by StickyThwomp** [comment]
> You don't really need to reply to each requested change when you fix it, it's okay to just resolve the conversation.

My bad, habit from work to be overly, painfully, annoyingly explicit regarding addressing everything :-D

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 11]
```diff
@@ -5,6 +5,8 @@
  */
 
 #include "z_en_nimotsu.h"
+#include "assets/objects/object_boj/object_boj.h"
+#include "macros.h"
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c:9]: The `macros.h` header should already be included by the `global.h` header, which is included by this actor header, so you can remove it safely

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 12]
```diff
@@ -28,25 +31,96 @@ const ActorInit En_Nimotsu_InitVars = {
...
+    position.y = this->actor.world.pos.y;
+    position.z = this->actor.world.pos.z + 2.0f;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Nimotsu/EnNimotsu_Update.s")
+    scale.x = 0.2f;
+    scale.y = 0.2f;
+    scale.z = 0.2f;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Nimotsu/EnNimotsu_Draw.s")
+    func_800BC620(&position, &scale, 0xFF, play);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c:124]: The third parameter is an alpha value, so use dec instead
```suggestion
    func_800BC620(&position, &scale, 255, play);
```

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 13]
```diff
@@ -28,25 +31,96 @@ const ActorInit En_Nimotsu_InitVars = {
...
+    EnNimotsu_UpdateCollision(this, play);
+}
 
-extern UNK_TYPE D_06013380;
+void EnNimotsu_Draw(Actor* thisx, PlayState* play) {
+    s32 pad;
+    EnNimotsu* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Nimotsu/func_80BE1C80.s")
+    Vec3f position;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c:107]: Please remove the extra empty line

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 14]
```diff
@@ -28,25 +31,96 @@ const ActorInit En_Nimotsu_InitVars = {
...
+    EnNimotsu* this = THIS;
+
+    Collider_DestroyCylinder(play, &this->collider);
+}
+
+void EnNimotsu_Update(Actor* thisx, PlayState* play) {
+    s32 pad;
+    EnNimotsu* this = THIS;
+
+    Vec3f dustPosition;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c:82]: Please remove the empty line
```suggestion
    EnNimotsu* this = THIS;
    Vec3f dustPosition;
```

**Review by AngheloAlf — APPROVED** [review]

**Review by AngheloAlf — CHANGES_REQUESTED** [review]
Wrong button, didn't intend to approve the PR yet :sweat_smile:

File: assets/xml/objects/object_boj.xml [hunk 15]
```diff
@@ -87,8 +87,8 @@
         <Animation Name="object_boj_Anim_011F84" Offset="0x11F84" />
         <Animation Name="object_boj_Anim_0128F4" Offset="0x128F4" />
         <Animation Name="object_boj_Anim_012E84" Offset="0x12E84" />
-        <DList Name="object_boj_DL_013380" Offset="0x13380" />
-        <Texture Name="object_boj_TLUT_0134C0" OutName="tlut_0134C0" Format="rgba16" Width="16" Height="16" Offset="0x134C0" />
-        <Texture Name="object_boj_Tex_0136C0" OutName="tex_0136C0" Format="ci8" Width="16" Height="16" Offset="0x136C0" />
+        <DList Name="gBombShopBagDL" Offset="0x13380" />
+        <Texture Name="gBombShopBagTLUT" OutName="bomb_shop_bag_tlut" Format="rgba16" Width="16" Height="16" Offset="0x134C0" />
+        <Texture Name="gBombShopBagTEX" OutName="bomb_shop_bag_tex" Format="ci8" Width="16" Height="16" Offset="0x136C0" />
```
Comment (engineer124) [inline assets/xml/objects/object_boj.xml:92]:
```suggestion
        <Texture Name="gBombShopBagTex" OutName="bomb_shop_bag" Format="ci8" Width="16" Height="16" Offset="0x136C0" />
```
Tex isn't usually all caps, and we leave `tex` out of the `OutName`

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.h [hunk 16]
```diff
@@ -7,8 +7,11 @@ struct EnNimotsu;
 
 typedef struct EnNimotsu {
     /* 0x000 */ Actor actor;
-    /* 0x144 */ char unk_144[0x98];
-} EnNimotsu; // size = 0x1DC
+    /* 0x144 */ UNK_TYPE1 unk_144[0x48];
+    /* 0x18C */ ColliderCylinder collider;
+    /* 0x1D8 */ u16 dustDone;
+    /* 0x1DA */ s16 dustTimer;
+} EnNimotsu; /* size = 0x1DC */
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.h:14]:
```suggestion
} EnNimotsu; // size = 0x1DC
```

**Review by engineer124 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c [hunk 17]
```diff
@@ -28,25 +30,94 @@ const ActorInit En_Nimotsu_InitVars = {
...
+
+void EnNimotsu_Update(Actor* thisx, PlayState* play) {
+    s32 pad;
+    EnNimotsu* this = THIS;
+    Vec3f dustPosition;
+
+    Actor_MoveWithGravity(&this->actor);
+
+    if (!(this->dustDone & 1) && (this->actor.bgCheckFlags & 1)) {
+
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.c:85]:
```suggestion
```

File: src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.h [hunk 18]
```diff
@@ -7,7 +7,10 @@ struct EnNimotsu;
 
 typedef struct EnNimotsu {
     /* 0x000 */ Actor actor;
-    /* 0x144 */ char unk_144[0x98];
+    /* 0x144 */ UNK_TYPE1 unk_144[0x48];
+    /* 0x18C */ ColliderCylinder collider;
+    /* 0x1D8 */ u16 dustDone;
+    /* 0x1DA */ s16 dustTimer;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Nimotsu/z_en_nimotsu.h:13]: There is only one timer, so `timer` will do.

**Review by AngheloAlf — APPROVED** [review]

**Review by EllipticEllipsis — APPROVED** [review]

**Pull request merged**